File: pshi.py
#!/usr/bin/env python3
# Crappy Pipeslash interpreter

# Copyright (c) 2017, SYZYGY-DEV333
# All rights reserved.
# Licensed under SPL 1.0 [splicense.pen.io]

import argparse
import re


# Parser

parser = argparse.ArgumentParser()
parser.add_argument("inputfile", help="input a pipeslash file")
args = parser.parse_args()

def file_to_string():
	with open(args.inputfile, 'r') as myfile:
		data=myfile.read().replace('\n', '')
	return data

infileraw = re.sub('[^/|\|]', '', file_to_string())
infile = re.sub(r'(?<!/)/(?!/)', '', infileraw).replace('//', ' ')

dic = {'|':'>',
	 '||':'<',
	 '|||':'+',
	 '||||':'-',
	 '|||||':'.',
	 '||||||':',',
	 '|||||||':'[',
	 '||||||||':']',
	 'sep':' ',
}

try: infile = infile.split(dic['sep'])
except ValueError: pass

outfile=''
for x in infile:
	if x in dic:
		outfile += dic[x]
	else:
		outfile += x


# Interpreter

def interpret(code):
  tape = [0]*(3*10**4)
  ptr = 0
  loops = []
  i = 0
  res = ""
  while i < len(code):
    c = code[i]
    #parse the operators, update the tape
    if c == "+":
      tape[ptr] = (tape[ptr] + 1) % 256
    elif c == "-":
      tape[ptr] = (tape[ptr] - 1) % 256
    elif c == ">":
      ptr += 1
      if ptr >= len(tape):
        raise ValueError('Segmentation fault')
    elif c == "<":
      ptr -= 1
      if ptr < 0:
        raise ValueError('Segmentation fault')
    elif c == ".":
      res += chr(tape[ptr])
    elif c == ",":
      c = ord(sys.stdin.read(1))
      if c != 26:
          tape[ptr] = c
    elif c == "[":
      #Check the tape
      if tape[ptr] > 0:
        #Add the index to the stack
        loops.append(i - 1)
      else:
        #Find the closing brace with a stack operation (get m to be 0)
        m = 1
        j = i + 1
        while j < len(code) and m > 0:
          if code[j] == "[":
            m += 1
          elif code[j] == "]":
            m -= 1
          j += 1
        if m > 0:
          raise ValueError('Mismatched loops')
        #move the loop
        i = j - 1
    elif c == "]":
      if len(loops) == 0:
        raise ValueError('Mismatched loops')
      #go back to the start of the loop
      i = loops.pop()
    i += 1
  return res

if __name__=="__main__":
  import sys
  print(interpret(outfile))

File: pshi.go
// pipeslash interpreter
// Copyright (c) 2017, SYZYGY-DEV333
// All rights reserved.
// Licensed under SPL 1.0 [bit.ly/splicense]
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strings"
)

var content_bytearray, err = ioutil.ReadFile(os.Args[1])
var content = string(content_bytearray[:])

func translate(src string) string {
	dic := strings.NewReplacer("/|/", ">",
		"/||/", "<",
		"/|||/", "+",
		"/||||/", "-",
		"/|||||/", ".",
		"/||||||/", ",",
		"/|||||||/", "[",
		"/||||||||/", "]",
		" ", "")
	return dic.Replace(src)
}

func interpret(src string) []uint8 {
	tape := []uint8{0}
	tapeIndex := 0

	srcLength := len(src)

	for srcIndex := 0; srcIndex < srcLength; srcIndex++ {

		switch src[srcIndex] {
		case '>':				// Move Pointer to the Right
			tapeIndex += 1
			if len(tape) <= tapeIndex {
				tape = append(tape, 0)
			}

		case '<':				// Move Pointer to the Left
			if tapeIndex > 0 {
				tapeIndex -= 1
			}

		case '+':				// Increment the memory cell under the pointer
			tape[tapeIndex] += 1

		case '-':				// Deincrement the memory cell under the pointer
			tape[tapeIndex] -= 1

		case '.':				// Output the character signified by the cell at the pointer
			fmt.Print(string(tape[tapeIndex]))

		case ',':				// Input a character and store it in the cell at the pointer
			b := make([]byte, 1)
			os.Stdin.Read(b)
			tape[tapeIndex] = b[0]

		case '[':			// jump past the matching `/||||||||/` if the cell under the pointer is 0
			if tape[tapeIndex] == 0 {
				for depth := 1; depth > 0; {
					srcIndex++
					srcCharacter := src[srcIndex]
					if srcCharacter == '[' {
						depth++
					} else if srcCharacter == ']' {
						depth--
					}
				}
			}

		case ']':			// jump back to the matching /|||||||/ if the cell under the pointer is nonzero
			for depth := 1; depth > 0; {
				srcIndex--
				srcCharacter := src[srcIndex]
				if srcCharacter == '[' {
					depth--
				} else if srcCharacter == ']' {
					depth++
				}
			}
			srcIndex--
		}
	}

	return tape
}

func main() {
	if err != nil {
		fmt.Println("Error reading file")
		fmt.Println(err)
	} else {
		var src = translate(content)
		interpret(src)
	}
}
